### ISSUE TITLE
CO2 conversion per MJ waste chp/power ccs should be flexible

### ISSUE THREAD

**Issue by MartLubben** [quintel/etmodel#3793]
**Situation:**
For NI we have added energy_chp_supercritical_ccs_waste_mix and energy_power_supercritical_ccs_waste_mix to the ETM. At this moment the molecule graph can read CO2 for the energy graph in two ways:

1. Per carrier in the energy node the demand can be multiplied by 'CO2 conversion per MJ'
2. Primary CO2 of an energy node can be queried. 

**Problem**
For some combined carriers (like network gas and waste mix) these options are not satisfactory. The first is not satisfactory because these combined carriers do not have a CO2 conversion per MJ. The second is not satisfactory because the primary co2 method only takes into account fossil CO2. So if a combined carrier partially consists of a bio carrier (e.g. network gas partially consisting of green gas) the on the spot emissions of green gas are not included. 

For network gas we 'solved' this by hard coding the CO2 emissions, see: energy_power_combined_cycle_ccs_network_gas_co2.ad. Regardless of the composition of network gas the on the spot CO2 is the same. 

For waste mix we cannot hard code because the on the spot emissions vary depending on the dataset. The mix of biogenic waste and non biogenic waste can be different for every dataset. 

**Solution:**
A solution would be to add a third method to calculate CO2 in the molecule graph. This third method could be to demand the sum of primary CO2 of an energy node including biogenic CO2. 

This could mean: primary_co2_emission + primary_co2_emission_of_bio_carriers

@antw Do you think this solution is feasible and do you have a time estimation?

Notifying @redekok and @michieldenhaan (cc @AlexanderWirtz)

**Comment by antw** [comment]
> This could mean: primary_co2_emission + primary_co2_emission_of_bio_carriers
>
> @antw Do you think this solution is feasible and do you have a time estimation?

This seems trivial. The hardest part will be coming up with a name for the new method. Any suggestions?

**Comment by MartLubben** [comment]
primary_co2_of_bio_and_fossil?

**Comment by antw** [comment]
> This could mean: primary_co2_emission + primary_co2_emission_of_bio_carriers

Just so that I can be sure that I understand: how does CCS and captured CO2 factor into this? I ask because what many think is "primary CO2 emission" in our model is really ["primary CO2 emission *minus* captured bio emissions"](https://github.com/quintel/etengine/blob/8838cb0f1a52e3082a41f7fa465685584bbb47d0/app/models/qernel/recursive_factor/primary_co2.rb#L9).

Are you sure it's correct to take primary demand – which has had inherited bio emissions subtracted – only to add all the bio emissions *including* that which would be captured?

**Comment by MartLubben** [comment]
Good question @antw . The use of this method is only for a query in the molecule graph to determine the CO2 emissions per MJ of the waste mix. I am not sure if this returns the right value. @michieldenhaan What do you think?

**Comment by michieldenhaan** [comment]
That's a good one.. Some background: What we would like to know is what the on-the-spot CO2 emissions of a node are, such that the molecule graph can show how much CO2 is emitted and how much is captured. If we have a node with e.g. `waste_mix` or `network_gas` as input part of the on-the-spot emissions will be fossil emissions and part will be bio emissions (depending on how much biogenic waste / green gas the carriers have). We cannot query `co2_conversion_per_mj` because these carriers do not have this attribute. We cannot query `weighted_co2_conversion_per_mj` because this does not include the bio emissions. So we need some sort of primary method that can trace how much of the waste mix / network gas is fossil and how much is bio and then add this up to give us the total emissions.

A downside of this solution is that primary is not always the same as on-the-spot. Suppose we have a waste mix burner with both waste mix and electricity as inputs. The CO2 emitted to produce the electricity is included in the primary method but those emissions do not take place at the location where the burner is located (so for the burner they are not on-the-spot emissions).

However, we only use the primary method in some very specific places in the molecule graph at the moment and in those cases getting the total primary demand is good enough. `primary_co2_emission + primary_co2_emission_of_bio_carriers`, so minus the captured bio emissions, will be fine because we currently don't use it in nodes for which the primary method passes a node with CCS.

My main concern is that this isn't very intuitive and that we might forget this in the future, resulting in incorrect outcomes. However, I'm afraid that a fix is not easy. This requires bookkeeping which emissions are emitted where and we currently don't do that.

So TL;DR: I guess it's fine to have a new method returning `primary_co2_emission + primary_co2_emission_of_bio_carriers` but we should be aware to use it carefully.

**Comment by MartLubben** [comment]
We could add a link to this issue in the node description.
